pkg/relay/ethereum/tx.go

```diff
@@ -504,107 +518,105 @@ func (iter *CallIter) Current() sdk.Msg {
 func (iter *CallIter) End() bool {
 	return len(iter.msgs) <= iter.cursor
 }
-func (iter *CallIter) Next() {
-	if !iter.End() {
-		iter.cursor += 1
+func (iter *CallIter) Next(n int) {
+	iter.cursor = min(len(iter.msgs), iter.cursor + n)
+}
+
+func (iter *CallIter) updateLoggerMessageInfo(logger *log.RelayLogger, from int, count int) {
+	if from < 0 || count < 0 || len(iter.msgs) <= from || len(iter.msgs) < from + count {
+		logger.Error("invalid parameter", fmt.Errorf("out of index: len(msgs)=%d, from=%d, count=%d", len(iter.msgs), from, count))
+		return
 	}
+
+	logger.Logger = logger.With(
+		logAttrMsgIndexFrom, from,
+		logAttrMsgCount, count,
+		logAttrMsgType, strings.Join(iter.msgTypeNames[0 : from + count], ","),
+	)
 }
 
-func (iter *CallIter) SendTx(ctx context.Context, c *Chain) (*gethtypes.Transaction, error) {
+func (iter *CallIter) BuildTx(ctx context.Context, c *Chain) (*gethtypes.Transaction, int, error) {
 	if c.multicall3 == nil {
-		return iter.sendSingleTx(ctx, c)
+		return iter.buildSingleTx(ctx, c)
 	} else {
-		return iter.sendMultiTx(ctx, c)
+		return iter.buildMultiTx(ctx, c)
 	}
 }
 
-func (iter *CallIter) sendSingleTx(ctx context.Context, c *Chain) (*gethtypes.Transaction, error) {
+func (iter *CallIter) buildSingleTx(ctx context.Context, c *Chain) (*gethtypes.Transaction, int, error) {
 	if iter.End() {
-		return nil, nil
+		return nil, 0, nil
 	}
 
 	logger := c.GetChainLogger()
-	logger = &log.RelayLogger{Logger: logger.With(
-		logAttrMsgIndexFrom, iter.Cursor(),
-		logAttrMsgType, fmt.Sprintf("%T", iter.Current()),
-	)}
+	iter.updateLoggerMessageInfo(logger, iter.Cursor(), 1)
 
 	opts, err := c.TxOpts(ctx, true);
 	if err != nil {
-		return nil, err
+		return nil, 0, err
 	}
 
 	// gas estimation
 	{
 		opts.GasLimit = math.MaxUint64
 		opts.NoSend = true
-		tx, err := c.SendTx(opts, iter.Current(), iter.skipUpdateClientCommitment)
+		tx, err := c.BuildMessageTx(opts, iter.Current(), iter.skipUpdateClientCommitment)
 		if err != nil {
 			logger.Error("failed to build tx for gas estimation", err)
-			return nil, err
+			return nil, 1, err
```

Comment:
The second return value of `buildSingleTx` sometimes becomes `0` and at other times takes on different values.
I couldn't understand the criteria for that.

Reply:
Its value used to inform parent caller to add log info.
When fail to build, if the error is caused by message, return message count=1 for logging, but the error is unrelated to message itself, returns message count=0.

Rewriting this behavior that logging not at caller but place of error occuring, not to return logging info.
And the function returns count only if buildTx is succeeded.